auth.php

```diff
@@ -323,10 +338,88 @@ protected function is_ip_restriction_enabled() {
      * @return object User object.
      */
     protected function create_user(array $data) {
-        // TODO:
-        // 1. Validate user
-        // 2. Create user.
-        // 3. Throw exception if something went wrong.
+        global $DB, $CFG;
+
+        $user = $data;
+        unset($user['ip']);
+        $user['auth'] = 'userkey';
+        $user['mnethostid'] = $CFG->mnet_localhost_id;
+
+        $requiredfieds = ['username', 'email', 'firstname', 'lastname'];
+        $missingfields = [];
+        foreach ($requiredfieds as $requiredfied) {
+            if (empty($user[$requiredfied])) {
+                $missingfields[] = $requiredfied;
+            }
+        }
+        if (!empty($missingfields)) {
+            throw new invalid_parameter_exception('Unable to create user, missing value(s): ' . implode(',', $missingfields));
+        }
+
+        if ($DB->record_exists('user', array('username' => $user['username'], 'mnethostid' => $CFG->mnet_localhost_id))) {
+            throw new invalid_parameter_exception('Username already exists: '.$user['username']);
+        }
+        if (!validate_email($user['email'])) {
+            throw new invalid_parameter_exception('Email address is invalid: '.$user['email']);
+        } else if (empty($CFG->allowaccountssameemail) &&
+            $DB->record_exists('user', array('email' => $user['email'], 'mnethostid' => $user['mnethostid']))) {
+            throw new invalid_parameter_exception('Email address already exists: '.$user['email']);
+        }
+
+        $userid = user_create_user($user);
+        return $DB->get_record('user', ['id' => $userid]);
+    }
+
+    /**
+     * Update an existing user.
+     *
+     * @param stdClass $user Existing user record.
+     * @param array $data Validated user data from web service.
+     *
+     * @return object User object.
+     */
+    protected function update_user(\stdClass $user, array $data) {
```

Comment:
I feel like this function may be refactored to make it easy to read, especially as you have a good test coverage.
But I don't have any specific recommendations for now. I'll let it through as it is and will refactor it later to make it more readable and generic. But if you want to refactor by yourself please let me know.